File: generated_code/cpp/solution_2079_3.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <map>
#include <algorithm>

using namespace std;

// Trie node structure representing a folder in the file system
struct TrieNode {
    string name;
    // Map to keep subfolders sorted by name to generate a unique, consistent subfolder structure string
    map<string, TrieNode*> children;
    bool deleted = false;

    TrieNode(string n = "") : name(n) {}
};

class Solution {
private:
    TrieNode* root;
    // Map to keep track of the frequency of each subfolder structure representation
    unordered_map<string, int> subfolderStructureCount;
    // Map to store all nodes associated with a specific subfolder structure representation
    unordered_map<string, vector<TrieNode*>> structureToNodes;

    // Helper to insert a path into the Trie
    void insertPath(const vector<string>& path) {
        TrieNode* curr = root;
        for (const string& folder : path) {
            if (curr->children.find(folder) == curr->children.end()) {
                curr->children[folder] = new TrieNode(folder);
            }
            curr = curr->children[folder];
        }
    }

    // Post-order traversal to serialize the subfolder structure of each node
    string serializeStructures(TrieNode* node) {
        if (!node) return "";
        
        string subStructure = "";
        // Process children in sorted order (guaranteed by std::map)
        for (auto& pair : node->children) {
            subStructure += "(" + pair.first + serializeStructures(pair.second) + ")";
        }

        // Only consider nodes that actually have subfolders for duplicate detection
        if (!subStructure.empty()) {
            subfolderStructureCount[subStructure]++;
            structureToNodes[subStructure].push_back(node);
        }

        return subStructure;
    }

    // DFS to extract all paths that haven't been marked for deletion
    void extractPaths(TrieNode* node, vector<string>& currentPath, vector<vector<string>>& result) {
        if (node->deleted) return;

        if (node != root) {
            currentPath.push_back(node->name);
            result.push_back(currentPath);
        }

        for (auto& pair : node->children) {
            extractPaths(pair.second, currentPath, result);
        }

        if (node != root) {
            currentPath.pop_back();
        }
    }

    // Helper to delete the dynamically allocated Trie memory
    void deleteTrie(TrieNode* node) {
        if (!node) return;
        for (auto& pair : node->children) {
            deleteTrie(pair.second);
        }
        delete node;
    }

public:
    vector<vector<string>> deleteDuplicateFolder(vector<vector<string>>& paths) {
        root = new TrieNode();

        // Step 1: Build the Trie from the input paths
        for (const auto& path : paths) {
            insertPath(path);
        }

        // Step 2: Serialize the structures and identify duplicates via post-order traversal
        serializeStructures(root);

        // Step 3: Mark nodes as deleted if their subfolder structure appears more than once
        for (auto& pair : subfolderStructureCount) {
            if (pair.second > 1) {
                for (TrieNode* node : structureToNodes[pair.first]) {
                    node->deleted = true;
                }
            }
        }

        // Step 4: Reconstruct the remaining non-deleted paths
        vector<vector<string>> result;
        vector<string> currentPath;
        extractPaths(root, currentPath, result);

        // Clean up memory to avoid leaks
        deleteTrie(root);

        return result;
    }
};

File: generated_code/python/solution_2079_3.py
from typing import List, Dict, Tuple

class TrieNode:
    def __init__(self, name: str = ""):
        self.name: str = name
        # Using a sorted dict (via sorting keys later) or a dict where we iterate in sorted order.
        # Since Python 3.7+ dicts preserve insertion order, we can sort keys when serializing.
        self.children: Dict[str, 'TrieNode'] = {}
        self.is_deleted: bool = False

class Solution:
    def deleteDuplicateFolder(self, paths: List[List[str]]) -> List[List[str]]:
        # Step 1: Build the Trie
        root = TrieNode()
        for path in paths:
            curr = root
            for folder in path:
                if folder not in curr.children:
                    curr.children[folder] = TrieNode(folder)
                curr = curr.children[folder]
                
        # Keep track of the structure serialized strings and the nodes that share them
        seen_structures: Dict[str, List[TrieNode]] = {}
        
        # Step 2: Serialize the subtree structure using Post-Order Traversal
        def serialize(node: TrieNode) -> str:
            if not node.children:
                return ""
            
            # Subfolders must be processed in a deterministic order (alphabetical)
            sub_serializations = []
            for child_name in sorted(node.children.keys()):
                child_str = serialize(node.children[child_name])
                sub_serializations.append(f"{child_name}({child_str})")
            
            # Combine the subfolder structures into a unique representation for the current node
            structure_str = ",".join(sub_serializations)
            
            # Register this structure to find duplicates
            if structure_str not in seen_structures:
                seen_structures[structure_str] = []
            seen_structures[structure_str].append(node)
            
            return structure_str
            
        serialize(root)
        
        # Step 3: Mark identical folders for deletion
        # Only folders with at least one subfolder (non-empty structure) can be duplicates
        for structure_str, nodes in seen_structures.items():
            if len(nodes) > 1:
                for node in nodes:
                    node.is_deleted = True
                    
        # Step 4: Construct the remaining paths using Pre-Order Traversal
        ans: List[List[str]] = []
        
        def collect_paths(node: TrieNode, current_path: List[str]) -> None:
            # If marked for deletion, skip this node and all of its subfolders
            if node.is_deleted:
                return
            
            # If it's not the root node, add its path to the final answer
            if node.name:
                current_path.append(node.name)
                ans.append(list(current_path))
                
            # Recurse for all children
            for child_name in node.children:
                collect_paths(node.children[child_name], current_path)
                
            # Backtrack
            if node.name:
                current_path.pop()
                
        collect_paths(root, [])
        return ans